callCrate.nix

```diff
@@ -0,0 +1,61 @@
+{ stdenv, crate2nix, callPackage, fetchurl, writeText, lib, runCommand, crates-io-index }:
+{ src
+, cargoLock ? src + "/Cargo.lock"
+, overrides ? { } }:
+let
+  inherit (builtins) filter match elemAt;
+
+  lock = builtins.fromTOML (builtins.readFile cargoLock);
+
+  parseMetadataKey = key:
+  let m = match "checksum ([A-Za-z0-9_-]*) ([A-Za-z0-9_.+-]*) \\(([A-Za-z0-9./_+:-]*)\\)" key;
+  in assert (! isNull m); {
+    name = elemAt m 0;
+    version = elemAt m 1;
+    source = elemAt m 2;
+  };
+
+  parsedMetadata = builtins.attrValues (builtins.mapAttrs (key: checksum: {
+    inherit (parseMetadataKey key) name version;
+    inherit checksum;
+  }) lock.metadata or {});
+
+  prefetch-crate = crate: fetchurl rec {
+    # We use .tar.gz here instead of .crate because this allows Nix to reuse this source for the actual build
+    name = "${crate.name}-${crate.version}.tar.gz";
+    url = "https://static.crates.io/crates/${crate.name}/${crate.name}-${crate.version}.crate";
+    sha256 = crate.checksum;
+  };
+
+  crates = map prefetch-crate (filter fetchable lock.package) ++ map prefetch-crate parsedMetadata;
+
+  fetchable = crate: crate ? checksum;
+
+  source-repository = runCommand "cargo-source-repository" { }
+  (''
+    mkdir $out
+    cp ${crates-io-index} -r $out/index
+  '' + lib.concatMapStringsSep "\n" (crate: "cp ${crate} $out/${builtins.replaceStrings [".tar.gz"] [".crate"] crate.name}") crates);
+
+  cargoConfig = writeText "config.toml" ''
+    [source.crates-io]
+    replace-with = "nix-crates-registry"
+
+    [source.nix-crates-registry]
+    local-registry = "${source-repository}"
```

Comment:
Won't we also want:

```toml
[net]
offline = false
```
?

https://doc.rust-lang.org/cargo/reference/config.html#netoffline

Reply:
I think we also need a way to add extra registries here. I'll probably do that next weekend.

Reply:
I think it will error our either way, but it might error out more helpfully this way...

Reply:
Actually, now that I think about it, crate2nix itself doesn't support non-crates.io registries yet, so it doesn't make sense to support them here. I'll just add `offline = true`.

Reply:
Done.